lib/xgu/xgu.h

```diff
@@ -0,0 +1,696 @@
+    };
+    struct {
+        float r, g, b, a;
+    };
+} XguVec4;
+
+typedef union {
+  float f[4*4];
+  XguVec4 col[4];
+} XguMatrix4x4;
+
+
+/* ========================= *
+ *  BEGIN PUSH HELPER BLOCK  *
+ * ========================= */
+
+static inline
+DWORD* push_command(DWORD* p, DWORD command, int nparams) {
+    pb_push(p++, command, nparams);
+    return p;
+}
+
+static inline
+DWORD* push_parameter(DWORD* p, DWORD parameter) {
+    *p++ = parameter;
+    return p;
+}
+
+static inline
+DWORD* push_boolean(DWORD* p, bool enabled) {
+    return push_parameter(p, enabled ? 1 : 0);
+}
+
+static inline
+DWORD* push_command_boolean(DWORD* p, DWORD command, bool enabled) {
+    p = push_command(p, command, 1);
+    p = push_boolean(p, enabled);
+    return p;
+}
+
+static inline
+DWORD* push_float(DWORD* p, float f) {
+    return push_parameter(p, *(uint32_t*)&f);
+}
+
+static inline
+DWORD* push_floats(DWORD* p, float* f, unsigned int count) {
+    for(unsigned int i = 0; i < count; i++) {
+        p = push_float(p, f[i]);
+    }
+    return p;
+}
+
+static inline
+DWORD* push_matrix2x2(DWORD* p, float m[2*2]) {
+    return push_floats(p, m, 2*2);
+}
+
+static inline
+DWORD* push_matrix4x4(DWORD* p, float m[4*4]) {
+    return push_floats(p, m, 4*4);
+}
+
+static inline
+DWORD* push_command_matrix2x2(DWORD* p, DWORD command, float m[2*2]) {
+    p = push_command(p, command, 2*2);
+    p = push_matrix2x2(p, m);
+    return p;
+}
+
+static inline
+DWORD* push_command_matrix4x4(DWORD* p, DWORD command, float m[4*4]) {
+    p = push_command(p, command, 4*4);
+    p = push_matrix4x4(p, m);
+    return p;
+}
+
+static inline
+DWORD* push_command_parameter(DWORD* p, DWORD command, DWORD parameter) {
+    p = push_command(p, command, 1);
+    p = push_parameter(p, parameter);
+    return p;
+}
+
+static inline
+DWORD* push_command_float(DWORD* p, DWORD command, float parameter) {
+    p = push_command(p, command, 1);
+    p = push_float(p, parameter);
+    return p;
+}
+
+
+/* ========================= *
+ * BEGIN XGU FUNCTIONS BLOCK *
+ * ========================= */
+
+inline
+DWORD* xgu_begin(DWORD* p, XguPrimitiveType type) {
+
+    // Force people to use xgu_end instead
+    assert(type != NV097_SET_BEGIN_END_OP_END);
+
+    return push_command_parameter(p, NV097_SET_BEGIN_END, type);
+}
+
+inline
+DWORD* xgu_end(DWORD* p) {
+    return push_command_parameter(p, NV097_SET_BEGIN_END, NV097_SET_BEGIN_END_OP_END);
+}
+
+inline
+DWORD* xgu_no_operation(DWORD* p, uint32_t param) {
+    /* param is typically 0 */
+    return push_command_parameter(p, NV097_NO_OPERATION, param);
+}
+
+inline
+DWORD* xgu_wait_for_idle(DWORD* p) {
+    return push_command(p, NV097_WAIT_FOR_IDLE, 0);
+}
+
+inline
+DWORD* xgu_set_viewport_offset(DWORD* p, float x, float y, float z, float w) {
+    p = push_command(p, NV097_SET_VIEWPORT_OFFSET, 4);
+    p = push_float(p, x);
+    p = push_float(p, y);
+    p = push_float(p, z);
+    return push_float(p, w);
```

Comment:
For code readability, we might want to use

```c
    p = push_float(p, y);
    p = push_float(p, z);
    p = push_float(p, w);
    return p;
```

Reply:
Fixed.